@here/harp-geoutils/lib/tiling/SubTiles.ts

```diff
@@ -7,78 +7,45 @@
-    private m_count: number;
-    private m_mask: number;
-    private m_shift: number;
-
-    constructor(tileKey: TileKey, level: number, mask: number) {
-        this.m_tileKey = tileKey;
-        this.m_level = level;
-        // tslint:disable:no-bitwise
-        this.m_count = 1 << (level << 1);
-        this.m_mask = mask;
-        this.m_shift = level > 2 ? (level - 2) << 1 : 0;
-        // tslint:enable:no-bitwise
-    }
-
-    get level(): number {
-        return this.m_level;
-    }
-    get tileKey(): TileKey {
-        return this.m_tileKey;
-    }
+    constructor(public tileKey: TileKey, public sizeX: number, public sizeY: number) {}
 
     [Symbol.iterator](): SubTiles.SubTileIterator {
-        return new SubTiles.SubTileIterator(this, this.m_count);
-    }
-
-    iterator() {
-        return this[Symbol.iterator]();
-    }
-
-    skip(index: number): number {
-        // tslint:disable:no-bitwise
-        if (this.m_mask !== ~0) {
-            while (index < this.m_count && (this.m_mask & (1 << (index >> this.m_shift))) === 0) {
-                ++index;
-            }
-        }
-        // tslint:enable:no-bitwise
-        return index;
+        return new SubTiles.SubTileIterator(this);
     }
 }
 
 export namespace SubTiles {
     export class SubTileIterator implements Iterator<TileKey> {
-        private m_parent: SubTiles;
         private m_index: number;
+        private m_parent: SubTiles;
         private m_totalSubTileCount: number;
 
-        constructor(parent: SubTiles, totalSubTileCount: number, index: number = 0) {
+        constructor(parent: SubTiles) {
+            this.m_index = -1;
             this.m_parent = parent;
-            this.m_index = parent.skip(index);
-            this.m_totalSubTileCount = totalSubTileCount;
+            this.m_totalSubTileCount = parent.sizeX * parent.sizeY;
         }
 
         get value() {
-            const parentKey = this.m_parent.tileKey;
-            const subLevel = this.m_parent.level;
+            const tk = this.m_parent.tileKey;
+            const dx = this.m_parent.sizeX;
+            const dy = this.m_parent.sizeY;
+            const py = Math.floor(this.m_index / dx);
+            const px = this.m_index % dx;
 
-            return TileKey.fromRowColumnLevel(
-                // tslint:disable:no-bitwise
-                (parentKey.row << subLevel) | (this.m_index >> subLevel),
-                (parentKey.column << subLevel) | (this.m_index & ((1 << subLevel) - 1)),
-                parentKey.level + subLevel
-                // tslint:enableno-bitwise
-            );
+            return TileKey.fromRowColumnLevel(tk.row * dy + py, tk.column * dx + px, tk.level + 1);
         }
 
-        next() {
-            const current = { value: this.value, done: this.m_index >= this.m_totalSubTileCount };
-            this.m_index = this.m_parent.skip(++this.m_index);
-            return current;
+        next(): IteratorResult<TileKey> {
```

Comment:
do you think you can refactor this code to use `yield`?

Reply:
thanks for the tip! the code is really simpler now.